rust/noosphere-gateway/src/middleware.rs

```diff
@@ -0,0 +1,244 @@
+        &self.sphere
+    }
+
+    pub fn sphere_mut(&mut self) -> &mut C {
+        &mut self.sphere
+    }
+
+    pub fn counterpart(&self) -> &Did {
+        &self.counterpart
+    }
+
+    pub async fn sphere_context(&self) -> Result<C::SphereContext> {
+        self.sphere.sphere_context().await
+    }
+
+    pub async fn sphere_context_mut(&mut self) -> Result<C::MutableSphereContext> {
+        self.sphere.sphere_context_mut().await
+    }
+}
+
+#[async_trait]
+impl<M, C, S, State> FromRequestParts<State> for GatewaySphereContext<M, C, S>
+where
+    State: Send + Sync,
+    M: ContextManager<C, S> + 'static,
+    C: HasMutableSphereContext<S>,
+    S: Storage + 'static,
+{
+    type Rejection = StatusCode;
+
+    async fn from_request_parts(parts: &mut Parts, state: &State) -> Result<Self, Self::Rejection> {
+        let context_manager = parts
+            .extensions
+            .get::<M>()
+            .ok_or_else(|| {
+                error!("Could not find ContextManager in extensions");
+                StatusCode::INTERNAL_SERVER_ERROR
+            })?
+            .clone();
+
+        // Read the counterpart from the hostname, which may not be a did:key,
+        // in the case of single-serve gateways in the cluster, or a self-hosted
+        // gateway without subdomain configuration.
+        // In the event of an invalid counterpart from hostname in single-serve settings,
+        // the context manager will still retrieve the sphere context, but we need to
+        // fetch the correct counterpart stored in the sphere DB.
+        let counterpart = get_counterpart_from_hostname(parts, state).await?;
+        let sphere = context_manager
+            .get_sphere_context(&counterpart)
+            .await
+            .map_err(map_bad_request)?;
+        let counterpart = {
+            let sphere_context = sphere.sphere_context().await.map_err(map_bad_request)?;
+            println!("Getting counterpart...");
+            sphere_context
+                .db()
+                .require_key(COUNTERPART)
+                .await
+                .map_err(map_bad_request)?
+        };
+        println!("counterpart...{}", counterpart);
+
+        Ok(GatewaySphereContext {
+            counterpart,
+            sphere,
+            marker_m: std::marker::PhantomData,
+            marker_c: std::marker::PhantomData,
+            marker_s: std::marker::PhantomData,
+        })
+    }
+}
+
+/// Represents the scope of a gateway request's authorization and sphere
+/// access.
+///
+/// Embodies the authorization status of the request-maker as it is
+/// represented by its `ucan` headers. Any request handler can use a [GatewayAuthority]
+/// to test if a required capability is satisfied by the authorization
+/// presented by the maker of the request, and provide access to the scoped
+/// sphere context.
+pub struct GatewayAuthority<M, C, S> {
+    proof_chain: ProofChain,
+    sphere: GatewaySphereContext<M, C, S>,
```

Comment:
I'm not sold on the `GatewayAuthority` "owning" the canonical `SphereContext` for the lifetime of the handler. It's not that it doesn't work, just that it seems like an unnecessary conflation of concepts without an obvious benefit.

Collapsing the number of arguments in a handler isn't a clear benefit to me because it obscures what is actually in use by the handler (and since Rust is verbose, it doesn't spare us from all the type bounds needed to express the shape of the `SphereContext`).

If it's about reducing work, I would suggest passing in the `SphereContext` from elsewhere at the time it is needed.

Otherwise, perhaps there is a join construct that pairs a `GatewayAuthority` with a `SphereContext`.

Are there other benefits to merging these concepts that I might be missing?

Reply:
Goal here is to reduce work -- IIRC, `FromRequestParts` is processed unordered, and if we need both authority and sphere context in a route, and the authority always needs a sphere context, then lets combine them. I agree that there's some conflating ideas here. There are a handful of ways of processing requests in axum, so maybe there's another one that fits the bill -- `Layers`, for example, is executed in order and does not take ownership IIRC. With functionality working currently, I can explore other options here (like `Layer`)

Reply:
Again, it's not that it doesn't work. My complaint is about the conceptual flattening happening here. The authority and the sphere context are two different concepts. If we were to make the claim that they are always delivered together, then I would suggest that there be a super construct to join them.

Reply:
Agreed on all fronts, new change addresses these well I think